src/common/attribute.rs

```diff
@@ -0,0 +1,626 @@
+		///
+		/// [`border_pixel`]: Attributes::border_pixel
+		const BORDER_PIXEL = 0x0000_0008;
+
+		/// See also: [`bit_gravity`], [`BitGravity`].
+		///
+		/// [`bit_gravity`]: Attributes::bit_gravity
+		const BIT_GRAVITY = 0x0000_0010;
+		/// See also: [`win_gravity`], [`WinGravity`].
+		///
+		/// [`win_gravity`]: Attributes::win_gravity
+		const WIN_GRAVITY = 0x0000_0020;
+
+		/// See also: [`backing_store`], [`BackingStore`].
+		///
+		/// [`backing_store`]: Attributes::backing_store
+		const BACKING_STORE = 0x0000_0040;
+		/// See also: [`backing_planes`], [`u32`].
+		///
+		/// [`backing_planes`]: Attributes::backing_planes
+		const BACKING_PLANES = 0x0000_0080;
+		/// See also: [`backing_pixel`], [`Pixel`].
+		///
+		/// [`backing_pixel`]: Attributes::backing_pixel
+		const BACKING_PIXEL = 0x0000_0100;
+
+		/// See also: [`override_redirect`], [`bool`].
+		///
+		/// [`override_redirect`]: Attributes::OverrideRedirect
+		const OVERRIDE_REDIRECT = 0x0000_0200;
+		/// See also: [`save_under`], [`bool`].
+		///
+		/// [`save_under`]: Attributes::save_under
+		const SAVE_UNDER = 0x0000_0400;
+
+		/// See also: [`event_mask`], [`EventMask`].
+		///
+		/// [`event_mask`]: Attributes::event_mask
+		const EVENT_MASK = 0x0000_0800;
+		/// See also: [`do_not_propagate_mask`], [`DeviceEventMask`].
+		///
+		/// [`do_not_propagate_mask`]: Attributes::do_not_propagate_mask
+		const DO_NOT_PROPAGATE_MASK = 0x0000_1000;
+
+		/// See also: [`colormap`], <code>[CopyableFromParent]<[Colormap]></code>.
+		///
+		/// [`colormap`]: Attributes::colormap
+		const COLORMAP = 0x0000_2000;
+
+		/// See also: [`cursor_appearance`], [`CursorAppearance`].
+		///
+		/// [`cursor_appearance`]: Attributes::cursor_appearance
+		const CURSOR_APPEARANCE = 0x0000_4000;
+	}
+}
+
+/// A boolean type wrapping a `u32` value for use in [`Attributes`].
+///
+/// This is used to easily encode booleans as four bytes. It is not part of the
+/// public API.
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, ConstantX11Size, X11Size, Readable, Writable)]
+struct Bool(u32);
+
+impl From<bool> for Bool {
+	fn from(value: bool) -> Self {
+		if value {
+			Self(1)
+		} else {
+			Self(0)
+		}
+	}
+}
+
+impl From<Bool> for bool {
+	fn from(value: Bool) -> Self {
+		value.0 != 0
+	}
+}
+
+/// A set of attributes for a [window].
+///
+/// [window]: Window
+#[derive(Debug, Hash, PartialEq, Eq)]
+pub struct Attributes {
+	/// Total [`X11Size`] of these `Attributes`.
+	///
+	/// This is cached so that it doesn't have to be recalculated each time -
+	/// `Attributes` is immutable.
+	///
+	/// This field is not part of the X11 format for this struct.
+	x11_size: usize,
+
+	mask: AttributeMask,
+
+	background_pixmap: Option<ParentRelatable<Option<Pixmap>>>,
+	background_pixel: Option<Pixel>,
+
+	border_pixmap: Option<CopyableFromParent<Pixmap>>,
+	border_pixel: Option<Pixel>,
+
+	bit_gravity: Option<BitGravity>,
+	win_gravity: Option<WinGravity>,
+
+	backing_store: Option<BackingStore>,
```

Comment:
These types (`BitGravity`, `WinGravity`, `BackingStore`) take up one byte each instead of 4.  They need to take up 4 bytes to fit as attributes.

Can likely use a similar solution as with the `Bool` type used for the Boolean fields here - use substitute types in the private fields, but convert them in the public API functions.  That means the private types can easily convert to `u32` format without affecting the public API at all.